Source/Particles/Collision/BinaryCollision/ElasticCollisionPerez.H

```diff
@@ -107,18 +107,48 @@ void ElasticCollisionPerez (
                amrex::max(n1,n2), T_R(-1.0/3.0) );
     lmdD = amrex::max(lmdD, rmin);
 
+#if (defined WARPX_DIM_RZ)
+    T_R * const AMREX_RESTRICT theta1 = soa_1.m_rdata[PIdx::theta];
+    T_R * const AMREX_RESTRICT theta2 = soa_2.m_rdata[PIdx::theta];
+#endif
+
     // call UpdateMomentumPerezElastic()
     {
       int i1 = I1s; int i2 = I2s;
       for (int k = 0; k < amrex::max(NI1,NI2); ++k)
       {
+
+#if (defined WARPX_DIM_RZ)
+          /* When particles are having the same vr,
+           * their velocities would change after collision,
+           * because they may have different azimuth angles,
+           * but their velocities are not supposed to be changed.
+           * To fix this, we rotate one particle to be in parallel with the other particle,
+           * such that their velocities won't change after collision if they have the same vr.
+           * Then, we rotate that particle back.
+           * Note that this only works for the lowest RZ mode, i.e., axisymmetry. */
+          AMREX_ALWAYS_ASSERT_WITH_MESSAGE(WarpX::n_rz_azimuthal_modes==1,
+            "RZ mode `warpx.n_rz_azimuthal_modes` must be 1 when using the binary collision module.");
+          T_R const theta = theta2[I2[i2]]-theta1[I1[i1]];
+          T_R const u1xbuf = u1x[I1[i1]];
+          u1x[I1[i1]] = u1xbuf*std::cos(theta) - u1y[I1[i1]]*std::sin(theta);
+          u1y[I1[i1]] = u1xbuf*std::sin(theta) + u1y[I1[i1]]*std::cos(theta);
+#endif
+
```

Comment:
Why write to the backing arrays rather than just using local variables for these two values? This seems like pointless memory traffic.

Reply:
Ok, so I see these are passed by reference, and updated inside `UpdateMomentumPerezElastic`. There's still an opportunity to do that write into a stack variable that won't be flushed to memory, and then only write it out once it's potentially been corrected by the following de-rotation.

Reply:
@PhilMiller Hi, thanks for you comments, could you provide more details how to improve the code, I'm not sure if I understand your point. Stack variables also work on GPUs?

Reply:
https://github.com/Yin-YinjianZhao/WarpX/pull/2
Here's a PR against your branch with my suggestion